[quant][graphmode] Enable inplace option for top level API

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#40414 [quant][graphmode] Enable inplace option for top level API**

Summary:
after `_reconstruct` is supported in RecursiveScriptModule: https://github.com/pytorch/pytorch/pull/39979
we can support inplace option in quantization API

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22178326](https://our.internmc.facebook.com/intern/diff/D22178326)

jerryzh168 committed:
[quant][graphmode] Enable inplace option for top level API

Summary:
after `_reconstruct` is supported in RecursiveScriptModule: https://github.com/pytorch/pytorch/pull/39979
we can support inplace option in quantization API

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Enable inplace option for top level API"


Summary:
after `_reconstruct` is supported in RecursiveScriptModule: https://github.com/pytorch/pytorch/pull/39979
we can support inplace option in quantization API

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22178326](https://our.internmc.facebook.com/intern/diff/D22178326)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Enable inplace option for top level API"


Summary:
after `_reconstruct` is supported in RecursiveScriptModule: https://github.com/pytorch/pytorch/pull/39979
we can support inplace option in quantization API

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22178326](https://our.internmc.facebook.com/intern/diff/D22178326)

[ghstack-poisoned]

vkuzo commented on test/quantization/test_quantize_jit.py:
```diff
@@ -1038,6 +1038,19 @@ def forward(self, x):
 
+    def test_inplace_option(self):
+        for tracing in [True, False]:
+            model = get_script_module(torch.nn.Conv2d(3, 3, 3), tracing, self.img_data[0][0])
```
I don't have too much context, but do we need to test (not in this PR, in the future) inplace on real world models to check the long tail of corner cases?  Or is this sufficient to prove that it works.

jerryzh168 replied:
I don't think we'll need to test `inplace` for all models, most of the models should just work. Although I'm not 100% sure if there will be cases that inplace does not work, maybe if we uncover such cases we can fix the problem add them to the test

raghuramank100 commented on torch/quantization/quantize_jit.py:
```diff
@@ -83,20 +97,17 @@ def convert_dynamic_jit(model, inplace=False, debug=False):
     else:
         assert run_fn, "Must provide calibration function for post training static quantization"
         assert run_args, "Must provide calibration dataset for post training static quantization"
         model = prepare_jit(model, qconfig_dict, inplace)
```
If inplace=True, why is prepare_jit returning a model? Should we have it not return anything if inplace=True?

jerryzh168 replied:
this is just a convenience for us that we return the model so that we don't need to have if branch here. Also this is the same as eager mode quantization API: https://github.com/pytorch/pytorch/blob/master/torch/quantization/quantize.py#L150

jerryzh168 committed:
Update on "[quant][graphmode] Enable inplace option for top level API"


Summary:
after `_reconstruct` is supported in RecursiveScriptModule: https://github.com/pytorch/pytorch/pull/39979
we can support inplace option in quantization API

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22178326](https://our.internmc.facebook.com/intern/diff/D22178326)

[ghstack-poisoned]
